### PR TITLE
*: use new row-format in tidb

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

using https://github.com/pingcap/tidb/blob/master/docs/design/2018-07-19-row-format.md format in TiDB

follow up https://github.com/pingcap/tidb/pull/7597 but get optimize for rowcodec from unionstore

tikv pr in https://github.com/tikv/tikv/pull/6132

### What is changed and how it works?

- TiDB:
  - make insert/update/replace encode new data as new-format(try to make encoder be reuse)
  - make point-get/batch-point-get decode new-format data into chunk directly(try to reuse decoder, but need more try..)
  - make batchcheck/ddl's get operation can decode with new-format(DecodeRowWithMapNew)
  - didn't modify cop relate logic, that has be resolved via arrow encode for chunk https://github.com/pingcap/tidb/pull/12536
- mocktikv:
   - make mocktikv read new-format from storage and back to old row-format to mocktikv's executor(mocktikv's executor still didn't using Chunk)
- unionstore: https://github.com/lysu/unistore/tree/new-row-format
   - make unionstore accept new-format update input(no need transform to old-format any more)
   - return new format for kv-get/ batch-get
- misc:
  - didn't change datum encode to keep statistic works(we don't follow design doc to change varint to int) https://github.com/ngaut/unistore/pull/237
  - keep binlog use old row format

WIP... need more tunning to make performance can be improved(now testing with unionstore).

### Check List 

Tests 

 - Unit test
 - Integration test(only integrate with unionstore or mocktikv, tikv wait development)
 - Manual test

```
Upgrade from old mocktikv or old unionstore to new format one, old table keep works
```
  
 - Performance test WIP 

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Breaking backward compatibility
 - Need take care upgrade step

Related changes

 - n/a

Release note

 - use new-row format in tidb

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/12634)

### CODE REVIEW

**Comment by lysu** [comment]
read performance by https://github.com/lysu/sysbench/tree/dev-add-garbage, using 170+ column table's point-get.

before:

```
sysbench --mysql-host=0.0.0.0 --mysql-port=4000 --mysql-user=root --threads=10 --tables=10 --table_size=30000 --time=60 --garbage_columns=170  oltp_point_select.lua run   
sysbench 1.1.0-1327e79 (using bundled LuaJIT 2.1.0-beta3)

Running the test with following options:
Number of threads: 10
Initializing random number generator from current time


Initializing worker threads...

Threads started!

SQL statistics:
    queries performed:
        read:                            673543
        write:                           0
        other:                           0
        total:                           673543
    transactions:                        673543 (11225.39 per sec.)
    queries:                             673543 (11225.39 per sec.)
    ignored errors:                      0      (0.00 per sec.)
    reconnects:                          0      (0.00 per sec.)

Throughput:
    events/s (eps):                      11225.3897
    time elapsed:                        60.0017s
    total number of events:              673543

Latency (ms):
         min:                                    0.15
         avg:                                    0.89
         max:                                   24.81
         95th percentile:                        2.14
         sum:                               599309.85

Threads fairness:
    events (avg/stddev):           67354.3000/84.41
    execution time (avg/stddev):   59.9310/0.00
```

after:

```
SQL statistics:
    queries performed:
        read:                            1091322
        write:                           0
        other:                           0
        total:                           1091322
    transactions:                        1091322 (18188.07 per sec.)
    queries:                             1091322 (18188.07 per sec.)
    ignored errors:                      0      (0.00 per sec.)
    reconnects:                          0      (0.00 per sec.)

Throughput:
    events/s (eps):                      18188.0701
    time elapsed:                        60.0021s
    total number of events:              1091322

Latency (ms):
         min:                                    0.11
         avg:                                    0.55
         max:                                   18.55
         95th percentile:                        1.06
         sum:                               598995.53

Threads fairness:
    events (avg/stddev):           109132.2000/64.02
    execution time (avg/stddev):   59.8996/0.00
```

**Comment by coocood** [comment]
@lysu 
how about `select_random_ranges.lua` workload?

**Comment by lysu** [comment]
we need take care upgrade step:

1. upgrade kv, tidb to new version but keep kv->db, db->kv protocol using old row format
2. open a switch in both side(or do another upgrade) to make kv->db, db->kv use new row format

**Comment by zhangjinpeng1987** [comment]
Amazing

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -1009,6 +1013,11 @@ func (s *SessionVars) SetSystemVar(name string, val string) error {
 		atomic.StoreUint64(&ExpensiveQueryTimeThreshold, uint64(tidbOptPositiveInt32(val, DefTiDBExpensiveQueryTimeThreshold)))
 	case TiDBTxnMode:
 		s.TxnMode = strings.ToUpper(val)
+	case TiDBRowFormatVersion:
+		formatVersion := int(tidbOptInt64(val, DefTiDBRowFormatVersionOldCluster))
+		if formatVersion != DefTiDBRowFormatVersionOldCluster {
```
Comment (coocood) [inline sessionctx/variable/session.go]: It's safer to check `formatVersion == DefTiDBRowFormatVersionNewCluster`

File: util/codec/codec.go [hunk 2]
```diff
@@ -176,7 +176,7 @@ func EncodeMySQLTime(sc *stmtctx.StatementContext, t types.Time, tp byte, b []by
 	if tp == mysql.TypeUnspecified {
 		tp = t.Type
 	}
-	if tp == mysql.TypeTimestamp && sc.TimeZone != time.UTC {
+	if tp == mysql.TypeTimestamp && sc != nil && sc.TimeZone != time.UTC {
```
Comment (coocood) [inline util/codec/codec.go]: sc can be nil?

File: tablecodec/tablecodec.go [hunk 3]
```diff
@@ -254,11 +255,27 @@ func EncodeValue(sc *stmtctx.StatementContext, b []byte, raw types.Datum) ([]byt
 	return codec.EncodeValue(sc, b, v)
 }
 
+// TODO: make this to a config.
+const useNewFormat = true
```
Comment (coocood) [inline tablecodec/tablecodec.go]: This can be removed?

File: executor/builder.go [hunk 4]
```diff
@@ -2490,16 +2491,51 @@ func (b *executorBuilder) buildSQLBindExec(v *plannercore.SQLBindPlan) Executor
...
+	return rowcodec.NewChunkDecoder(reqCols, handleColID, defVal, ctx.GetSessionVars().TimeZone)
+}
+
 func (b *executorBuilder) buildBatchPointGet(plan *plannercore.BatchPointGetPlan) Executor {
 	startTS, err := b.getStartTS()
 	if err != nil {
 		b.err = err
 		return nil
 	}
+	decoder := newRowDecoder(b.ctx, plan.Schema(), plan.TblInfo)
```
Comment (zhangjinpeng1987) [inline executor/builder.go]: Can we reuse the Decoder for the same table?

File: tablecodec/tablecodec.go (hunk 3 above)
Comment (jackysp) [inline tablecodec/tablecodec.go]: could it be removed?

File: util/rowDecoder/decoder_test.go [hunk 5]
```diff
@@ -14,6 +14,7 @@
 package decoder
 
 import (
+	"github.com/pingcap/tidb/util/rowcodec"
```
Comment (jackysp) [inline util/rowDecoder/decoder_test.go]: format

File: util/codec/codec.go (hunk 2 above)
Comment (lysu) [reply util/codec/codec.go]: in one of previous commit, sc maybe nil, but in current it should never be nil, remove this change.

File: executor/builder.go (hunk 4 above)
Comment (lysu) [reply executor/builder.go]: Our `PointGetExecutor` is pointget + selection , so `plan.Schema()` will get different results for `select a from t where id = 1` and `select a, b from t where id = 1`, in previous commits, I have try to reuse decoder in cached plan level, but after discuss with others, it introduce some complex code, maybe we can renew at here but do it later?

**Comment by lysu** [comment]
/bench +tpcc
```
workload:
  prepare: true
```

**Comment by coocood** [comment]
LGTM

**Pull request merged**